syntax = "proto3";

package api.services.v1;
option go_package = "github.com/ScienceObjectsDB/go-api/api/services/v1";
option java_multiple_files = true;
option java_package = "com.github.ScienceObjectsDB.java_api.api.services.v1";
option java_outer_classname = "DatasetServices";


import "api/services/v1/dataset_service_models.proto";
import "google/api/annotations.proto";

// Dataset management service
// Manages all dataset related services
// All data objects are associated with one data dataset
// Dataset versions group these data objects, which makes them reusable
service DatasetService {
    // CreateNewDataset Creates a new dataset and associates it with a dataset
    rpc CreateDataset(CreateDatasetRequest) returns (CreateDatasetResponse) {
        option (google.api.http) = {
            post: "/api/v1/dataset/create"
            body: "*"
        };
    }

    // Dataset Returns a specific dataset
    rpc GetDataset(GetDatasetRequest) returns (GetDatasetResponse) {
        option (google.api.http) = {
            post: "/api/v1/dataset/get"
            body: "*"
        };
    }

    // Lists Versions of a dataset
    rpc GetDatasetVersions(GetDatasetVersionsRequest) returns (GetDatasetVersionsResponse) {
        option (google.api.http) = {
            post: "/api/v1/datasetversions/list"
            body: "*"
        };
    }

    rpc GetDatasetObjectGroups(GetDatasetObjectGroupsRequest) returns (GetDatasetObjectGroupsResponse) {
        option (google.api.http) = {
            post: "/api/v1/dataset/list"
            body: "*"
        };
    }

    rpc GetCurrentObjectGroupRevisions(GetCurrentObjectGroupRevisionsRequest) returns (GetCurrentObjectGroupRevisionsResponse) {
        option (google.api.http) = {
            get: "/api/v1/dataset/{id}/currentgroupversions"
        };
    }

    // Updates a field of a dataset
    rpc UpdateDatasetField(UpdateDatasetFieldRequest) returns (UpdateDatasetFieldResponse) {
        option (google.api.http) = {
            post: "/api/v1/dataset/update"
            body: "*"
        };
    }

    // DeleteDataset Delete a dataset
    rpc DeleteDataset(DeleteDatasetRequest) returns (DeleteDatasetResponse) {
        option (google.api.http) = {
            delete: "/api/v1/dataset/{id}"
        };
    }

    //---------------------------------------------------------------------------------------
    // Dataset version calls

    //ReleaseDatasetVersion Release a new dataset version
    rpc ReleaseDatasetVersion(ReleaseDatasetVersionRequest) returns (ReleaseDatasetVersionResponse) {
        option (google.api.http) = {
            post: "/api/v1/datasetversion/create"
            body: "*"
        };
    }

    rpc GetDatasetVersion(GetDatasetVersionRequest) returns (GetDatasetVersionResponse) {
        option (google.api.http) = {
            post: "/api/v1/datasetversion/get"
            body: "*"
        };
    }

    rpc GetDatsetVersionRevisions(GetDatsetVersionRevisionsRequest) returns (GetDatsetVersionRevisionsResponse) {
        option (google.api.http) = {
            post: "/api/v1/datasetversion/list"
            body: "*"
        };
    }

    rpc DeleteDatasetVersion(DeleteDatasetVersionRequest) returns (DeleteDatasetVersionResponse) {
        option (google.api.http) = {
            delete: "/api/v1/datasetversion/{id}"
        };
    }
}